<a href="https://colab.research.google.com/github/AireesRondain/PM/blob/main/PM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install mlxtend --upgrade
import pandas as pd

In [ ]:
from google.colab import drive
drive.mount('/gdrive')
%cd /gdrive/My Drive/Colab Notebooks/paymaya

Drive already mounted at /gdrive; to attempt to forcibly remount, call drive.mount("/gdrive", force_remount=True).
/gdrive/My Drive/Colab Notebooks/paymaya


In [ ]:
df = pd.read_csv('Data Analyst_Assessment_v2.csv')

In [ ]:
df.head()

customer_number          use_case  ... spend_volume  spend_transaction_count
0         10000003  Pay with PayMaya  ...      3865.63                        2
1         10000003     Bills Payment  ...       836.00                        1
2         10000003      PAYMAYA SHOP  ...      3562.00                       13
3         10000010                QR  ...       214.00                        1
4         10000010   Online Purchase  ...       549.00                        1

[5 rows x 6 columns]

In [ ]:
df.tail()

customer_number          use_case  ... spend_volume  spend_transaction_count
2495         10024476       P2P Sending  ...        900.0                        1
2496          1002448      PAYMAYA SHOP  ...        100.0                        1
2497         10024501  Pay with PayMaya  ...       1399.0                        1
2498         10024501      PAYMAYA SHOP  ...       1206.0                       13
2499         10024516      PAYMAYA SHOP  ...        200.0                        1

[5 rows x 6 columns]

In [ ]:
#convert customer_number to string
df['customer_number'] = df['customer_number'].astype(str) 

In [ ]:
#dataframe to build the recommendation system
txns = df[['customer_number', 'use_case']]
txns.head()

customer_number          use_case
0        10000003  Pay with PayMaya
1        10000003     Bills Payment
2        10000003      PAYMAYA SHOP
3        10000010                QR
4        10000010   Online Purchase

Show each customer with a list of use_cases availed

In [ ]:
def string_list(x):
  return [str(i) for i in x]

txns = txns.groupby('customer_number')['use_case'].apply(list).reset_index()
txns.head()

customer_number                                           use_case
0        10000003    [Pay with PayMaya, Bills Payment, PAYMAYA SHOP]
1        10000010  [QR, Online Purchase, PAYMAYA SHOP, P2P Sendin...
2        10000039                                 [Pay with PayMaya]
3        10000057                                 [Instapay Sending]
4        10000067                                     [PAYMAYA SHOP]

In [ ]:
uc_group = df.groupby('customer_number')['use_case'].apply(list).reset_index()
uc_group["num_use_case"] = kyc_group['use_case'].apply(lambda x: len(x))
uc_group.head()

customer_number  ... num_use_case
0        10000003  ...            3
1        10000010  ...            5
2        10000039  ...            1
3        10000057  ...            1
4        10000067  ...            1

[5 rows x 3 columns]

In [ ]:
uc_group.tail()

customer_number  ... num_use_case
1254        10024451  ...            2
1255        10024476  ...            4
1256         1002448  ...            1
1257        10024501  ...            2
1258        10024516  ...            1

[5 rows x 3 columns]

In [ ]:
uc_group.sort_values('num_use_case',ascending=False)

customer_number  ... num_use_case
68          10001240  ...            8
448         10008786  ...            7
1169        10022749  ...            7
26          10000689  ...            7
344         10006577  ...            7
...              ...  ...          ...
596         10011432  ...            1
598         10011457  ...            1
600         10011463  ...            1
601         10011473  ...            1
1258        10024516  ...            1

[1259 rows x 3 columns]

In [ ]:
#convert customer_number to string
uc_group['customer_number'] = uc_group['customer_number'].astype(str) 

In [ ]:
left_join = pd.merge(df,uc_group, on = 'customer_number', how='left')

In [ ]:
left_join

customer_number  ... num_use_case
0           10000003  ...            3
1           10000003  ...            3
2           10000003  ...            3
3           10000010  ...            5
4           10000010  ...            5
...              ...  ...          ...
2495        10024476  ...            4
2496         1002448  ...            1
2497        10024501  ...            2
2498        10024501  ...            2
2499        10024516  ...            1

[2500 rows x 8 columns]

In [ ]:
#save to csv
left_join.to_csv('paymaya_dataset.csv', index=False)

In [ ]:
from mlxtend.preprocessing import TransactionEncoder

In [ ]:
te = TransactionEncoder()

te.fit(txns['use_case'])
txns_1hot = te.transform(txns['use_case'])

In [ ]:
txns_1hot = pd.DataFrame(txns_1hot, columns = te.columns_)
txns_1hot.head()

ATM Withdrawal  Bills Payment  ...  Pay with PayMaya     QR
0           False           True  ...              True  False
1           False          False  ...             False   True
2           False          False  ...              True  False
3           False          False  ...             False  False
4           False          False  ...             False  False

[5 rows x 10 columns]

In [ ]:
txns_1hot.tail()

ATM Withdrawal  Bills Payment  ...  Pay with PayMaya     QR
1254           False           True  ...             False  False
1255           False          False  ...             False  False
1256           False          False  ...             False  False
1257           False          False  ...              True  False
1258           False          False  ...             False  False

[5 rows x 10 columns]

10 columns x 1258 rows (total)

Using FP Growth (Frequency-Pattern Growth) Algorithm

In [ ]:
!pip install fp-growth

In [ ]:
%%timeit -n1 -r1

fpgrowth(txns_1hot, min_support=0.05, max_len=3, use_colnames=True)

1 loop, best of 1: 14.9 ms per loop


In [ ]:
is_fp = fpgrowth(txns_1hot, min_support=0.05, max_len=5, use_colnames=True)
is_fp.sort_values('support',ascending=False)

support                                         itemsets
0   0.537728                                   (PAYMAYA SHOP)
3   0.344718                               (Instapay Sending)
1   0.274027                                  (Bills Payment)
4   0.209690                                (Online Purchase)
5   0.204925                                    (P2P Sending)
2   0.190627                               (Pay with PayMaya)
15  0.186656                 (PAYMAYA SHOP, Instapay Sending)
9   0.152502                    (Bills Payment, PAYMAYA SHOP)
10  0.110405                (Bills Payment, Instapay Sending)
19  0.104051                      (PAYMAYA SHOP, P2P Sending)
16  0.101668                  (PAYMAYA SHOP, Online Purchase)
12  0.088165                 (PAYMAYA SHOP, Pay with PayMaya)
20  0.086577                  (Instapay Sending, P2P Sending)
6   0.084194                                             (QR)
17  0.081811              (Online Purchase, Instapay Sending)
11  0.073074  (Bills Payment, PAYMAYA SHOP, Instapay Sending)
7   0.065925              (P2P - Consumer to Negosyo Sending)
8   0.060365                                 (ATM Withdrawal)
18  0.058777                 (Bills Payment, Online Purchase)
21  0.058777                     (Bills Payment, P2P Sending)
22  0.058777    (PAYMAYA SHOP, Instapay Sending, P2P Sending)
13  0.051628                (Bills Payment, Pay with PayMaya)
14  0.050834             (Pay with PayMaya, Instapay Sending)

##Calculate Association Rules

In [ ]:
from mlxtend.frequent_patterns import association_rules

In [ ]:
rules = association_rules(is_fp, metric="lift", min_threshold=1.1)
rules

antecedents  ... conviction
0                     (Bills Payment)  ...   1.097438
1                  (Instapay Sending)  ...   1.068042
2       (Bills Payment, PAYMAYA SHOP)  ...   1.258141
3   (Bills Payment, Instapay Sending)  ...   1.367144
4    (PAYMAYA SHOP, Instapay Sending)  ...   1.193033
5                     (Bills Payment)  ...   1.109105
6                      (PAYMAYA SHOP)  ...   1.029497
7                  (Instapay Sending)  ...   1.075480
8                   (Online Purchase)  ...   1.074500
9                  (Instapay Sending)  ...   1.036237
10                 (Instapay Sending)  ...   1.061732
11                      (P2P Sending)  ...   1.134649
12   (PAYMAYA SHOP, Instapay Sending)  ...   1.160514
13        (PAYMAYA SHOP, P2P Sending)  ...   1.505999
14    (Instapay Sending, P2P Sending)  ...   1.439646
15                     (PAYMAYA SHOP)  ...   1.025518
16                 (Instapay Sending)  ...   1.080116
17                      (P2P Sending)  ...   1.140450

[18 rows x 9 columns]

In [ ]:
rules["antecedent_len"] = rules["antecedents"].apply(lambda x: len(x))
rules

antecedents  ... antecedent_len
0                     (Bills Payment)  ...              1
1                  (Instapay Sending)  ...              1
2       (Bills Payment, PAYMAYA SHOP)  ...              2
3   (Bills Payment, Instapay Sending)  ...              2
4    (PAYMAYA SHOP, Instapay Sending)  ...              2
5                     (Bills Payment)  ...              1
6                      (PAYMAYA SHOP)  ...              1
7                  (Instapay Sending)  ...              1
8                   (Online Purchase)  ...              1
9                  (Instapay Sending)  ...              1
10                 (Instapay Sending)  ...              1
11                      (P2P Sending)  ...              1
12   (PAYMAYA SHOP, Instapay Sending)  ...              2
13        (PAYMAYA SHOP, P2P Sending)  ...              2
14    (Instapay Sending, P2P Sending)  ...              2
15                     (PAYMAYA SHOP)  ...              1
16                 (Instapay Sending)  ...              1
17                      (P2P Sending)  ...              1

[18 rows x 10 columns]

In [ ]:
#save to csv
df.to_csv('paymaya_lift.csv', index=False)

In [ ]:
# If I'm interested in the rules that satisfy the ff: criteria:
rules[ (rules['antecedent_len'] >= 1) &
       (rules['confidence'] > 0.60) &
       (rules['lift'] > 1.1) ]

# 66% of the time if (bills payment + Instapay sending) -> Paymaya Shop

antecedents  ... antecedent_len
3   (Bills Payment, Instapay Sending)  ...              2
14    (Instapay Sending, P2P Sending)  ...              2

[2 rows x 10 columns]

In [ ]:
rules[rules['antecedents'] == {'PAYMAYA SHOP'}]

antecedents  ... antecedent_len
6   (PAYMAYA SHOP)  ...              1
15  (PAYMAYA SHOP)  ...              1

[2 rows x 10 columns]